app/src/main/java/com/google/android/sambadocumentsprovider/ShareManager.java

```diff
@@ -64,56 +66,119 @@
     // Loading mounted servers
     final Set<String> serverStringSet =
         mPref.getStringSet(SERVER_STRING_SET_KEY, Collections.<String> emptySet());
-    final Map<String, CredentialTuple> credentialMap = new HashMap<>(serverStringSet.size());
+    final Map<String, ShareTuple> shareMap = new HashMap<>(serverStringSet.size());
     for (String serverString : serverStringSet) {
       // TODO: Add decryption
-      String uri = decode(serverString, credentialMap);
+      String uri = decode(serverString, shareMap);
       if (uri != null) {
         mServerStringMap.put(uri, serverString);
       }
     }
 
     mServerStringSet = new HashSet<>(serverStringSet);
 
-    for (Map.Entry<String, CredentialTuple> server : credentialMap.entrySet()) {
-      final CredentialTuple tuple = server.getValue();
+    for (Map.Entry<String, ShareTuple> server : shareMap.entrySet()) {
+      final ShareTuple tuple = server.getValue();
+
+      if (tuple.mIsMounted) {
+        mMountedServerSet.add(server.getKey());
+      }
+
       mCredentialCache.putCredential(
           server.getKey(), tuple.mWorkgroup, tuple.mUsername, tuple.mPassword);
     }
   }
 
-  public synchronized void mountServer(
-      String uri, String workgroup, String username, String password, ShareMountChecker checker)
-      throws IOException {
-    if (mServerStringMap.containsKey(uri)) {
-      throw new IllegalStateException("Uri " + uri + " is already mounted.");
+  /**
+   * Stores the server info (e.g. uri, credentials) permanently.
+   * @param mount the share will be displayed as a separate root.
+   * @param force if the share with such a uri already exists, it will be overwritten.
+   */
+  public synchronized void addServer(
+          String uri, String workgroup, String username, String password,
+          ShareMountChecker checker, boolean mount, boolean force) throws IOException {
+    if (!force && mServerStringMap.containsKey(uri)) {
```

Comment:
Force is prone to be abused because it loosens a check. Having such parameter will make devs think they can pass wrong value to it to get around a sanity check.

If you need addOrUpdateServer, just change this method to it and remove the check.